backend/api/EventHandlers/MissionScheduling.cs

```diff
@@ -0,0 +1,343 @@
+                }
+            }
+
+            robot.CurrentMissionId = null;
+            await _robotService.Update(robot);
+        }
+
+        public async Task ScheduleMissionToReturnToSafePosition(string robotId, string areaId)
+        {
+            var area = await _areaService.ReadById(areaId);
+            if (area == null)
+            {
+                _logger.LogError("Could not find area with ID {AreaId}", areaId);
+                return;
+            }
+            var robot = await _robotService.ReadById(robotId);
+            if (robot == null)
+            {
+                _logger.LogError("Robot with ID: {RobotId} was not found in the database", robotId);
+                return;
+            }
+            var closestSafePosition = ClosestSafePosition(robot.Pose, area.SafePositions);
+            // Cloning to avoid tracking same object
+            var clonedPose = ObjectCopier.Clone(closestSafePosition);
+            var customTaskQuery = new CustomTaskQuery
+            {
+                RobotPose = clonedPose,
+                Inspections = new List<CustomInspectionQuery>(),
+                InspectionTarget = new Position(),
+                TaskOrder = 0
+            };
+
+            var missionRun = new MissionRun
+            {
+                Name = "Drive to Safe Position",
+                Robot = robot,
+                MissionRunPriority = MissionRunPriority.Emergency,
+                InstallationCode = area.Installation!.InstallationCode,
+                Area = area,
+                Status = MissionStatus.Pending,
+                DesiredStartTime = DateTimeOffset.UtcNow,
+                Tasks = new List<MissionTask>(new[]
+                {
+                    new MissionTask(customTaskQuery)
+                }),
+                Map = new MapMetadata()
+            };
+
+            await _missionRunService.Create(missionRun);
+        }
+
+        public static bool MissionRunQueueIsEmpty(IList<MissionRun> missionRunQueue)
+        {
+            return !missionRunQueue.Any();
+        }
+        private async void OnIsarUnavailable(Robot robot)
+        {
+            robot.Enabled = false;
+            robot.Status = RobotStatus.Offline;
+            if (robot.CurrentMissionId != null)
+            {
+                var missionRun = await _missionRunService.ReadById(robot.CurrentMissionId);
+                if (missionRun != null)
+                {
+                    missionRun.SetToFailed();
+                    await _missionRunService.Update(missionRun);
+                    _logger.LogWarning(
+                        "Mission '{Id}' failed because ISAR could not be reached",
+                        missionRun.Id
+                    );
+                }
+            }
+            robot.CurrentMissionId = null;
+            await _robotService.Update(robot);
+        }
+
+        public static Pose ClosestSafePosition(Pose robotPose, IList<SafePosition> safePositions)
```

Comment:
We have quite a few supporting functions in this file. Should we split out a utility file for these type of functions?

Reply:
We can't change the ones that update the database to another file because that caused us to run into issues before, so it would be creating a file for only the last two functions, I believe. This file was originally much smaller but we had to change it. I can try to move those two and as many as possible from the other tho, if you think it's worth it

Reply:
This doesn't lead to problems if the mission queue is still frozen after clearing the emergency state when the robot gets available because no emergency state mission will be found so the robot will just keep the first mission in the queue and run it, that's why it works the way it has been implemented